### PR TITLE
Fixes #7407 - added tests to make sure things are working as expected

### PR DESCRIPTION
**Pull request by unorthodoxgeek**

### CODE REVIEW

**Comment by unorthodoxgeek** [comment]
[test]

File: app/models/hostgroup.rb [hunk 1]
```diff
@@ -195,9 +195,15 @@ def clone(name = "")
     new.puppetclasses = puppetclasses
     new.locations     = locations
     new.organizations = organizations
+    new.config_groups = self.config_groups
```
Comment (dLobatog) [inline app/models/hostgroup.rb]: No need for the `self`, keep it consistent with the values above

File: app/models/hostgroup.rb [hunk 2]
```diff
@@ -195,9 +195,15 @@ def clone(name = "")
...
+
     # Clone any parameters as well
     self.group_parameters.each{|param| new.group_parameters << param.clone}
-    self.config_groups.each{|group| new.config_groups << group}
+    self.lookup_values.each do |lookup_value|
+      new_lookup_value = lookup_value.dup
+      new_lookup_value.match = "hostgroup=#{name}"
```
Comment (domcleal) [inline app/models/hostgroup.rb]: title?

**Comment by domcleal** [comment]
Use 'fixes' in the commit message as it changes behaviour and will resolve the ticket.

File: app/models/hostgroup.rb (hunk 2 above)
Comment (unorthodoxgeek) [inline app/models/hostgroup.rb]: don't understand what you mean...

File: app/models/concerns/host_common.rb [hunk 3]
```diff
@@ -30,7 +30,7 @@ module HostCommon
 
     alias_method :all_puppetclasses, :classes
 
-    has_many :lookup_values, :finder_sql => Proc.new { LookupValue.where('lookup_values.match' => lookup_value_match).to_sql }, :dependent => :destroy
+    has_many :lookup_values, :finder_sql => Proc.new { LookupValue.where('lookup_values.match' => lookup_value_match).to_sql }, :dependent => :destroy, :validate => false
```
Comment (dLobatog) [inline app/models/concerns/host_common.rb:33]: Why did you remove the validation?

File: app/models/hostgroup.rb (hunk 2 above)
Comment (domcleal) [inline app/models/hostgroup.rb]: Given a nested host group "Foo/Bar", its name is "Bar" but its title is "Foo/Bar".  I think the matcher needs to use the title.

File: app/models/concerns/host_common.rb (hunk 3 above)
Comment (unorthodoxgeek) [inline app/models/concerns/host_common.rb:33]: because if the hostgroup gets saved, its lookup values are validated before it, and are still invalid (because there's no such hostgroup)

File: test/unit/hostgroup_test.rb [hunk 4]
```diff
@@ -357,14 +357,41 @@ def setup_user(operation)
...
+      group.save
+      cloned = group.clone("new_name")
+      assert cloned.puppetclasses.size == 1
+      assert cloned.puppetclasses == group.puppetclasses
+    end
+
+    test "clone should clone parameters" do
+      group.group_parameters.create!(:name => "foo", :value => "bar")
+      cloned = group.clone("new_name")
+      cloned.save!(:validate => false)
```
Comment (dLobatog) [inline test/unit/hostgroup_test.rb:381]: No need to save and hit the database here, `cloned` contains `group_parameters` already at this point.

File: test/unit/hostgroup_test.rb [hunk 5]
```diff
@@ -357,14 +357,41 @@ def setup_user(operation)
...
+      assert cloned.group_parameters.map{|p| [p.name, p.value]} == group.group_parameters.map{|p| [p.name, p.value]}
```
Comment (dLobatog) [inline test/unit/hostgroup_test.rb]: `assert_equal` , `.map` is not necessary, you can just compare `.group_parameters`

File: test/unit/hostgroup_test.rb [hunk 6]
```diff
@@ -357,14 +357,41 @@ def setup_user(operation)
...
+    end
+
+    test "clone should clone lookup values" do
+      lv = lookup_values(:four)
+      lv.match = group.send(:lookup_value_match)
+      lv.save!
+      cloned = group.clone("new_name")
+      cloned.save
+      assert cloned.lookup_values.map(&:value) == group.lookup_values.map(&:value)
```
Comment (dLobatog) [inline test/unit/hostgroup_test.rb]: Ditto as in the previous test

File: test/unit/hostgroup_test.rb [hunk 7]
```diff
@@ -357,14 +357,41 @@ def setup_user(operation)
...
-
-    cloned = group.clone("new_name")
-    assert cloned.config_groups.include?(config_group)
+  context "#clone" do
+    let(:group) {FactoryGirl.create(:hostgroup, :name => 'a')}
+
+    test "clone should clone config groups as well" do
+      config_group = ConfigGroup.create!(:name => 'Blah')
+      group.config_groups << config_group
+      group.save
```
Comment (dLobatog) [inline test/unit/hostgroup_test.rb]: `.save` the group or the cloned group is not needed in any of the tests

File: app/models/concerns/host_common.rb (hunk 3 above)
Comment (dLobatog) [inline app/models/concerns/host_common.rb:33]: Hmmm... for some reason tests don't indicate that, I remove the validation and lookup values are cloned properly. I'm also concerned about `Host::Managed` lookup_values here.  Could you write a test that shows what you are saying in that comment?

File: test/unit/hostgroup_test.rb (hunk 7 above)
Comment (unorthodoxgeek) [inline test/unit/hostgroup_test.rb]: yes it is. removed it to check, #clone goes to the db in some of the fields, and in the cloned group we REALLY need the save, because we need to make sure everything actually gets saved to the db and is retrievable.

File: test/unit/hostgroup_test.rb (hunk 5 above)
Comment (unorthodoxgeek) [inline test/unit/hostgroup_test.rb]: comparing group parameters is done by identity, i.e. the item's ID column, not by their values. thus, yeah, map is needed. I'll change the assertion to assert_equal.

File: test/unit/hostgroup_test.rb (hunk 4 above)
Comment (unorthodoxgeek) [inline test/unit/hostgroup_test.rb:381]: group parameters are cloned, i.e. new records. we want to make sure they're saved to the DB

**Comment by unorthodoxgeek** [comment]
[test]

File: test/unit/hostgroup_test.rb (hunk 7 above)
Comment (dLobatog) [inline test/unit/hostgroup_test.rb]: Then add more tests if you want to ensure that behavior (although I think it's much better as it is now, you're testing clone, not validations), currently tests will pass if you remove all the `.save` calls in this context group (except for the lookup_value).

**Comment by dLobatog** [comment]
@unorthodoxgeek See inline comments, tests are fine as they are now (clone is tested) in my opinion, no need to check they are saved in the db. Also the two commits have to be rebased into one with 'fixes' as @domcleal said above, thanks!

File: app/models/concerns/host_common.rb (hunk 3 above)
Comment (unorthodoxgeek) [inline app/models/concerns/host_common.rb:33]: don't understand what you mean. the test fails if you keep validation on that association, by definition I've written a test that shows what I'm saying here.

**Comment by dLobatog** [comment]
Merged as 7691ab4a56a49118f6cf97ca7f6de1c932b99e2a, thanks @unorthodoxgeek
